Appsilon/rhino
# Add version switcher to the site

### Changes
Closes #530

This script provides a function for building a pkgdown site with multiple versions.

Versions are defined by:
  - ref - a git ref to build. In the gist, prefix tags with `refs/tags/` and branches with `refs/remotes/origin/`
  - href - a URL path where the version will be available at
  - label - a label to display in a version switcher. Using true will take version from DESCRIPTION and append "(dev)" suffix

The script works as follows:

1. Copy repository into a temporary directory
2. Read `pkgdown/navbar_template.html` and populate switcher options with versions
3. For every version[^1]:
    a. Create a git worktree based on version's ref
    b. Populate the navbar template with current version's label
    c. Write the navbar template to `pkgdown/templates/navbar.html` relative to the git worktree
    d. Copy `pkgdown/extra.css` from the repository we are building out of (this ensures a consistent and up-to-date styling)
    e. Use pkgdown to build site
    f. Remove the worktree

[^1]: Version with "/" href will be built first. This avoids an error from pkgdown regarding building the site to a non-empty directory (that's not managed by pkgdown).

The resulting build has the site with "/" href in the destination directory. Respective versions are stored in subdirectories named after labels defining versions.

#### Known issues
  1. Every version has its own `sitemap.xml`
  1. Every version has separate assets (Bootstrap, favicons, ...)
  1. All versions are indexed by search engines
  1. None of the versions has canonical URLs
  1. "Source" link in vignettes and function reference is hardcoded to HEAD (instead of a respective version)
  1. Switching a version redirects to index (instead of staying on the same subpage in target version if it exists)
  1. Rhinoverse button resembling a back button makes is even more confusing. Pressing it neither puts you back to the default version nor to a root of the current version you are viewing. It sends you to an entirely new website.

To address 1.-4. it's best to contribute directly to pkgdown. I think the effort of adjusting this script and keeping relevant templates up-to-date would be bigger than implementing a proper version support directly in pkgdown.

Additionally, I pinned bslib version to keep Bootstrap 5.2.2. Using latest bslib that provides Bootstrap 5.3.0 resulted in styling partially breaking on the site. However, the styling altogether should be revisited and improved.

### How to test
1. Visit https://tymekdev.github.io/rhino
3. Use the switcher in the navbar
4. Observe how content and URL changes after selecting a new version
